presences/src/main/java/fr/openent/presences/helper/SquashHelper.java

```diff
@@ -24,46 +24,85 @@ public SquashHelper() {
-            courses.addAll(splitCoursesEvent);
-        }
+    public static List<Course> squash(List<Course> coursesEvent, List<Course> splitCoursesEvent,
+                                      JsonArray registerEvent, boolean multipleSlot) {
+
+        List<Course> courses = new ArrayList<>();
         JsonObject registers = groupRegisters(registerEvent);
+
+        setCoursesRegistersInfos(coursesEvent, registers, false);
+        setCoursesRegistersInfos(splitCoursesEvent, registers, true);
+
+
+        List<Course> createdCourses = coursesEvent.stream()
+                .filter(course -> course.getRegisterId() != null).collect(Collectors.toList());
+
+        List<Course> createdSplitCourses = splitCoursesEvent.stream()
+                .filter(course -> course.getRegisterId() != null).collect(Collectors.toList());
+
+
+        courses.addAll(filterDuplicateCourses(createdSplitCourses, createdCourses));
+        courses.addAll(filterDuplicateCourses(createdCourses, createdSplitCourses));
+
+
+        List<Course> futureSplitCourses = splitCoursesEvent.stream()
+                .filter(fSplitCourse -> fSplitCourse.getRegisterId() == null
+                        && coursesEvent.stream()
+                        .noneMatch(course -> course.getRegisterId() != null &&
+                                Objects.equals(course.getId(), fSplitCourse.getId())))
+                .collect(Collectors.toList());
+
+
+        List<Course> futureCourses = coursesEvent.stream()
+                .filter(fCourse -> fCourse.getRegisterId() == null
+                        && splitCoursesEvent.stream()
+                        .noneMatch(course -> course.getRegisterId() != null &&
+                                Objects.equals(course.getId(), fCourse.getId())))
+                .collect(Collectors.toList());
+
+        courses.addAll(Boolean.TRUE.equals(multipleSlot) ? futureSplitCourses : futureCourses);
+
+        return courses;
+    }
+
+    private static List<Course> filterDuplicateCourses(List<Course> toFilter, List<Course> toCheck) {
+        return toFilter.stream()
+                .filter(c1 -> toCheck.stream()
+                        .noneMatch(c2 -> Objects.equals(c1.getId(), c2.getId())))
+                .collect(Collectors.toList());
+    }
+
+
+    private static void setCoursesRegistersInfos(List<Course> courses, JsonObject registers, boolean isSplitSlot) {
         for (Course course : courses) {
-            boolean found = false;
-            int j = 0;
+            course.setSplitSlot(isSplitSlot);
+
             JsonArray courseRegisters = registers.getJsonArray(course.getId());
             if (courseRegisters == null) {
                 continue;
             }
-            while (!found && j < courseRegisters.size()) {
-                JsonObject register = courseRegisters.getJsonObject(j);
+
+            for (int i = 0; i < courseRegisters.size(); i++) {
+                JsonObject register = courseRegisters.getJsonObject(i);
```

Comment:
List<JsonObject> registers = courseRegisters.getList(); (avec le p'tit suppress warning qui va bien)

ça permet d'éviter cette ligne pas hyper jolie: JsonObject register = courseRegisters.getJsonObject(i);
Et en plus de ça, tu peux le stream pour abstraire l'itération et rendre la lecture du code plus lisible.
(ça c'est pas obligatoire !)